### PR TITLE
Enhance compliments remote file with refresh support

### PR DESCRIPTION
**Pull request by sdetweil**
add support to refresh the compliments remotefile
add testcases for both without refresh (testcase missing) and with refresh

doc to be updated

### CODE REVIEW

**Comment by sdetweil** [comment]
runner step check needs a fix if  branch is changed to develop it doesn't get it, rest of steps correct

File: modules/default/compliments/compliments.js [hunk 1]
```diff
@@ -1,4 +1,8 @@
 /* global Cron */
+/* global global */
+
+console.log("window name='"+window.name+"'")
```
Comment (khassel) [inline modules/default/compliments/compliments.js:4]: left from debugging?

File: modules/default/compliments/compliments.js [hunk 2]
```diff
@@ -1,4 +1,8 @@
...
+const compliments_test_mode=(window.name=='jsdom')?true:false
```
Comment (khassel) [inline modules/default/compliments/compliments.js:5]: defined but not used

File: modules/default/compliments/compliments.js [hunk 3]
```diff
@@ -236,6 +258,25 @@ Module.register("compliments", {
...
+			// use them
+			if(JSON.stringify(this.config.compliments)!== JSON.stringify(this.compliments_new)){
+				// only reset if the contents changes
+				this.config.compliments = this.compliments_new
+				// reset the index
+				this.lastIndexUsed = -1;
+			}
+		}
+		// in test mode only
+		//if (compliments_test_mode) {
```
Comment (khassel) [inline modules/default/compliments/compliments.js:273]: as mentioned above defined but not used here

**Comment by khassel** [comment]
@rejas have we a problem with linting?

This PR contains stuff where I would expect linting errors but there are none.

**Comment by sdetweil** [comment]
i left code in so we could resolve the problem

File: modules/default/compliments/compliments.js (hunk 1 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:4]: yes for debugging why window.name is empty in test run mode

File: modules/default/compliments/compliments.js (hunk 2 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:5]: the wanted testcase to run
you see my two approaches, both fail

**Comment by sdetweil** [comment]
speaking of lint problems.

if I run  test:js or lint:js I see this 
```text
npm run test:js

> magicmirror@2.30.0-develop test:js
> eslint .

/home/sam/MagicMirror/js/es.js
  15:7  error  Parsing error: 'return' outside of function

/home/sam/MagicMirror/tests/configs/port_variable.js
  3:1  error  Parsing error: Unexpected token }
```

File: modules/default/compliments/compliments.js (hunk 3 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:273]: again this was left to help resolve the window.name problem

**Comment by sdetweil** [comment]
oops.. implemented minimum delay before intest #3631 incorporated

**Comment by sdetweil** [comment]
added documentation update

**Comment by rejas** [comment]
I merged your doc update already, what is to be done here @sdetweil ?

**Comment by rejas** [comment]
> @rejas have we a problem with linting?
> 
> This PR contains stuff where I would expect linting errors but there are none.

I dont get any errors anymore, so probably fixed?

**Comment by sdetweil** [comment]
i need to change the code here to match the test mode detection in #3631

**Comment by sdetweil** [comment]
fails waiting for #3631

**Comment by sdetweil** [comment]
ready for review

**Comment by khassel** [comment]
> I dont get any errors anymore, so probably fixed?

@rejas not fixed, an example:

```diff
diff --git a/modules/default/compliments/compliments.js b/modules/default/compliments/compliments.js
index 7b4be902..b76713fb 100644
--- a/modules/default/compliments/compliments.js
+++ b/modules/default/compliments/compliments.js
@@ -20,24 +20,25 @@ Module.register("compliments", {
                afternoonStartTime: 12,
                afternoonEndTime: 17,
                random: true,
-               specialDayUnique: false,
+               specialDayUnique: false
        },
-       urlSuffix:"",
-       compliments_new:null,
-       refreshMinimumDelay:15*60*60*1000,  // 15 minutes
+       urlSuffix: "",
+       compliments_new: null,
+       refreshMinimumDelay: 15 * 60 * 60 * 1000, // 15 minutes
        lastIndexUsed: -1,
        // Set currentweather from module
```

**Comment by sdetweil** [comment]
and the two i posted above

npm run test:js

File: modules/default/compliments/compliments.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 /* global Cron */
+/* global global */
```
Comment (KristjanESPERANTO) [inline modules/default/compliments/compliments.js:2]: Why did you add this line? 'global' is already defined as a built-in global variable.

Comment (sdetweil) [reply modules/default/compliments/compliments.js:2]: was for trying to make earlier test global.window.name == 'jsdom' work... will remove
(as global was undefined)

**Comment by khassel** [comment]
@sdetweil can you please merge `develop` into this (or rebase) and fix the (upcoming) lint errors? Thanks!

**Comment by sdetweil** [comment]
it was built on develop...  i just didn't pick develop on submission

**Comment by khassel** [comment]
`develop` was updated, we need the newest version here to get correct linting

**Comment by sdetweil** [comment]
yes, updated, repushed

**Comment by sdetweil** [comment]
lint broken now

**Comment by KristjanESPERANTO** [comment]
> lint broken now

Most of the issues should be fixable by `npm run lint:js`.

**Comment by sdetweil** [comment]
not ..

now, I have personal files in the MM folders that don't pass the lint tests and its none of the linters business..
but they are rejected
```
> magicmirror@2.30.0-develop lint:js
> eslint . --fix


/home/sam/MagicMirror.old/js/es.js
  15:7  error  Parsing error: 'return' outside of function

/home/sam/MagicMirror.old/modules/default/calendar/save/calendarfetcherutils.js
  120:2   warning  Method 'filterEvents' has too many lines (518). Maximum allowed is 400     max-lines-per-function
  145:32  warning  Arrow function has too many lines (486). Maximum allowed is 400            max-lines-per-function
  299:23  error    Expected '!==' and instead saw '!='                                        eqeqeq
  306:19  error    Expected '!==' and instead saw '!='                                        eqeqeq
  325:10  error    Unexpected constant condition                                              no-constant-condition
  336:19  error    Expected '!==' and instead saw '!='                                        eqeqeq
  349:10  error    Unexpected constant condition                                              no-constant-condition
  349:10  error    Unexpected constant truthiness on the left-hand side of a `&&` expression  no-constant-binary-expression
  415:59  error    Unexpected comma in middle of array                                        no-sparse-arrays
  443:40  error    Expected '!==' and instead saw '!='                                        eqeqeq
  488:19  error    Expected '!==' and instead saw '!='                                        eqeqeq
  490:31  error    Expected '!==' and instead saw '!='                                        eqeqeq
  497:32  error    Expected '===' and instead saw '=='                                        eqeqeq
  507:34  error    Expected '===' and instead saw '=='                                        eqeqeq
  524:11  error    Expected no linebreak before this statement                                @stylistic/nonblock-statement-body-position
  527:11  error    Expected no linebreak before this statement                                @stylistic/nonblock-statement-body-position
  527:23  error    'eventDiff' is assigned to itself                                          no-self-assign

/home/sam/MagicMirror.old/modules/default/calendar/testrrule.js
  9:38  error  Parsing error: Invalid number

/home/sam/MagicMirror.old/modules/default/textit.js
  20:17  error  Parsing error: Unexpected token function

/home/sam/MagicMirror.old/tests/configs/port_variable.js
  3:1  error  Parsing error: Unexpected token }

```

leave me alone!

I think THIS is a lint error

this.config.remoteFileRefreshInterval !== 0 
forcing !== for a numeric  value, there is no ambiguity

**Review by khassel — APPROVED** [review]

**Comment by khassel** [comment]
> leave me alone!

I knew you would like it ;)
